8044 CE consolidated waitlist support

ttoomey opened the pull request:
## _Merging this PR_
- use the squash-merge strategy for PRs targeting a release-X branch

#### Major Changes

* **UnitGroup-Based Matching**: Waitlist rules and candidate pools are now managed at the `UnitGroup` level instead of individual `Opportunity` records. 

* **Real-Time Pool Updates**: Removed the periodic `BuildCandidatePoolsJob`. Rule and UnitGroup changes now trigger candidate pool rebuilds through ActiveRecord callbacks.

* **Required CandidatePool Setup**: Units can only be marked available for CE if they are associated with a UnitGroup. The `markUnitsAvailable` mutation enforces this requirement.

* **Rule Snapshots**: Referrals show eligibility criteria from when the Opportunity was created, using the `assignment_rules` on the opportunity record.

#### Implementation Details

* **Rule Ownership**: Match rules are limited to `UnitGroup`, `Project`, or `Organization` owners; rules can no longer be owned by `Unit` or `Opportunity`
* **Schema Changes**: Added `candidate_pool_id` foreign key to `hmis_unit_groups` table
* **Code Restructure**: Split `CandidatePoolResolver` into:
  * `UnitGroupRuleResolver`: Determines the effective set of rules for a `UnitGroup`
  * `CandidatePoolRepository`: Handles CandidatePool persistence and lookup
* **Rule Precedence**: Deterministic ordering (`UnitGroup > Project > Organization`, then by ID)
* **Data Integrity**: `Rule` and `UnitGroup` models prevent changes to critical attributes after creation
* **Service Refactor**: `CandidatePoolBuilder` is now stateless; callers handle locks and transactions
* **Cleanup Updates**: Orphan cleanup checks both `UnitGroup` and `Opportunity` references; `ProcessClientsJob` removes dangling `ChangeMarker` records

## Type of change
New feature

## Checklist before requesting review
[//]: # (Remove any items that are not applicable)
- [x] I have performed a self-review of my code
- [x] I have run the code that is being changed under ideal conditions, and it doesn't fail
- [x] I have updated the documentation (or not applicable)
- [x] I have added spec tests (or not applicable)
- [ ] I have provided testing instructions in this PR or the related issue (or not applicable)

ttoomey committed:
8044-unit-group-pools-squashed

ttoomey committed:
fixes

ttoomey committed:
docs, fixes

ttoomey committed:
docs

ttoomey committed:
cleanups

ttoomey committed:
refactor

ttoomey committed:
tidy method name

ttoomey committed:
test fixes

ttoomey committed:
fixes

ttoomey committed:
fixes

ttoomey committed:
docs, fix

ttoomey commented:
@gigxz I think this is ready for an initial review.

ttoomey committed:
docs

gigxz reviewed (commented):
This looks great!! 
I can re-review and test after https://github.com/greenriver/hmis-warehouse/pull/5632 is merged in,  but overall this looks great I don't have any concerns. Thanks for adding the consistency bit for rule sorting and priority scheme selection.

ttoomey committed:
Merge branch 'release-176' into 8044-unit-group-pools-squashed

ttoomey committed:
mark avail doesn't need to dirty pools

ttoomey committed:
support opportunity creation when pools are not yet built

ttoomey committed:
docs

ttoomey committed:
processing, dirty marking, misc

ttoomey committed:
dirty marker cleanup

ttoomey committed:
Merge branch 'release-176' into 8044-unit-group-pools-squashed

ttoomey committed:
test fixes

ttoomey commented on app/models/grda_warehouse/tasks/client_cleanup.rb:
```diff
@@ -10,6 +10,36 @@
 # also set the processed_hash on the enrollment to nil
 
 module GrdaWarehouse::Tasks
+  # The ClientCleanup task is a maintenance utility for ensuring the  integrity and
```
@eanders - I generated some docs for this class while adding the hooks for CE. Look okay?

eanders replied:
Looks, great, thank you!

ttoomey commented on drivers/hmis/app/graphql/mutations/ce/mark_units_available.rb:
```diff
@@ -14,56 +14,48 @@ class Ce::MarkUnitsAvailable < CleanBaseMutation
-        end
-
-        result = Hmis::Ce::Opportunity.import!(opportunities)
-        Hmis::Ce::BuildCandidatePoolsJob.perform_later(opportunity_ids: result.ids)
```
I don't think this is necessary now, opportunities don't impact pool generation

ttoomey committed:
docs

ttoomey committed:
fixes

ttoomey committed:
fixes

ttoomey committed:
tests

ttoomey commented on drivers/hmis/app/models/hmis/ce/README_FOR_CE_PROCESSING.md:
```diff
@@ -0,0 +1,137 @@
+# Coordinated Entry (CE) Change Tracking System
```
renamed from `README_FOR_CHANGE_MARKER.md`

ttoomey commented:
Note, this also addresses some issues that came up on dev stand up:
* add change tracking to client clean up routine
* avoid creating pools if either requirement or priority is null
* pool builder no longer polls continuously, it only re-enqueues if there are more pools to process

gigxz commented on drivers/hmis/app/models/hmis/ce/README_FOR_CE_PROCESSING.md:
```diff
@@ -0,0 +1,137 @@
+
+### Event Triggers and Actions
+
+The following table summarizes the key events that trigger actions within the CE system:
```
Very helpful format! Thank you!

eanders commented on drivers/hmis/app/models/hmis/ce/match/candidate_pool_builder.rb:
```diff
@@ -1,117 +1,149 @@
+      update_stale_flags!
+      cleanup_orphan_pools
+
+      Rails.logger.info(
```
Is there a reason this isn't using the `log_info` helper?

ttoomey replied:
good catch, thanks!

eanders commented on drivers/hmis/app/models/hmis/ce/match/README_FOR_CE_MATCH_ENGINE.md:
```diff
@@ -28,3 +28,16 @@ The classes within this module are organized into several subdirectories to grou
+## Concepts and Contracts (Quick Reference)
+
+- **Candidate Pool Key**: `[priority_expression, requirement_expression]` (Strings). Default/no‑specific‑rules returns `nil` (no pool is created; associations remain `NULL`).
+- **Rule Precedence**: Priority scheme selection is by owner precedence `UnitGroup > Project > Organization`; tie-breaker by rule `id`. Eligibility requirements from all applicable owners are combined with logical AND in deterministic order.
```
Not sure this is the right place, but somewhere we should note that per-unit rules should be implemented by putting the units in individual unit groups.  This was our work-around that allows us to streamline the places candidate pools are attached.

gigxz committed:
Propose API fix for resolving unit rules

gigxz committed:
Only if latest oppt is active

gigxz reviewed (approved):
Looks great! it needs one fix for the Unit page, here is my proposed and manually tested fix: https://github.com/greenriver/hmis-warehouse/pull/5645

Other than that, looks good to merge! nice work.

ttoomey committed:
Merge pull request #5645 from greenriver/gig/8044-unit-group-pools-squashed-api-fix

Propose API fix for resolving unit rules

ttoomey committed:
Merge branch 'release-176' into 8044-unit-group-pools-squashed

ttoomey committed:
pr feedback

ttoomey committed:
cleanup, tests

ttoomey committed:
Merge branch 'release-177' into 8044-unit-group-pools-squashed

ttoomey committed:
spec labels

ttoomey commented:
> it needs one fix for the Unit page, here is my proposed and manually tested fix: https://github.com/greenriver/hmis-warehouse/pull/5645

Thanks! I merged and added test coverage for it (see drivers/hmis/spec/requests/hmis/get_units_spec.rb). Small adjustment so that unit.eligibilityRequirements always returns an array (for simplicity).

ttoomey commented:
@gigxz I'm going to retarget this to 177

gigxz reviewed (approved):
sounds good, thanks!
